### PR TITLE
Add python bindings for FeatureModelBuilder

### PR DESCRIPTION
**Pull request by sarvech76**

### CODE REVIEW

**Comment by sarvech76** [comment]
-cleaned the repository
-updated featureModelBuilder python bindings

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+#include "pybind11/detail/common.h"
+#include "pybind11/pybind11.h"
+#include <pybind11/stl.h>
+#include <pybind11/complex.h>
+#include <pybind11/functional.h>
+#include <pybind11/chrono.h>
+#include <filesystem>
+#include <iostream>
+#include <bits/stdc++.h>
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
#include "vara/Feature/FeatureModel.h"
#include "vara/Feature/FeatureModelBuilder.h"
#include "vara/Feature/FeatureModelTransaction.h"

#include "pybind11/pybind11.h"
#include "pybind11/stl.h"
#include "pybind_common.h"

#include <filesystem>
#include <memory>
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            [](vf::FeatureModelBuilder &FMB, std::string name){
+                return FMB.makeRoot(name);
+            }    
+        )
+        .def(
+            "build_feature_model",
+            [](vf::FeatureModelBuilder &FMB){
+                return FMB.buildFeatureModel();
+            }    
+        );
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "build_feature_model",
            [](vf::FeatureModelBuilder &FMB){
                return FMB.buildFeatureModel();
            }
        );
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    py::class_<vf::FeatureModelBuilder>(M,  "FeatureModelBuilder")
+        .def(py::init([](){
+                      return vf::FeatureModelBuilder();
+                      }))
+        .def(
+            "make_binary_feature",
+            [](vf::FeatureModelBuilder &FMB, std::string feature_name, bool Opt){
+                return FMB.makeFeature<vf::BinaryFeature>(feature_name); 
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "make_binary_feature",
            [](vf::FeatureModelBuilder &FMB, std::string feature_name, bool Opt){
                return FMB.makeFeature<vf::BinaryFeature>(feature_name); 
            }
        )
```

Please don't add any trailing whitespaces.

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            [](vf::FeatureModelBuilder &FMB, std::string feature_name, bool Opt){
+                return FMB.makeFeature<vf::BinaryFeature>(feature_name); 
+            }    
+        )   
+        .def(
+            "add_an_edge",
+            [](vf::FeatureModelBuilder &FMB, std::string parent_name, std::string feature_name){
+                return FMB.addEdge(parent_name, feature_name);
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "add_an_edge",
            [](vf::FeatureModelBuilder &FMB, std::string parent_name, std::string feature_name){
                return FMB.addEdge(parent_name, feature_name);
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                if (relation == "OR"){
+                    RK = vf::Relationship::RelationshipKind::RK_OR;
+                }
+                else if (relation == "XOR"){
+                    RK = vf::Relationship::RelationshipKind::RK_ALTERNATIVE;
+                }
+                
+                return FMB.emplaceRelationship(RK, parent_name);
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "emplace_relationship",
            [](vf::FeatureModelBuilder &FMB, std::string parent_name, std::string relation) {
 
                vf::Relationship::RelationshipKind RK;
                if (relation == "OR"){
                    RK = vf::Relationship::RelationshipKind::RK_OR;
                }
                else if (relation == "XOR"){
                    RK = vf::Relationship::RelationshipKind::RK_ALTERNATIVE;
                }

                return FMB.emplaceRelationship(RK, parent_name);
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        .def(
+            "add_boolean_constraint",
+            [](vf::FeatureModelBuilder &FMB, vf::Constraint &C){
+                return FMB.addConstraint(std::make_unique<vf::FeatureModel::BooleanConstraint> (C.clone()) );
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "add_boolean_constraint",
            [](vf::FeatureModelBuilder &FMB, vf::Constraint &C){
                return FMB.addConstraint(std::make_unique<vf::FeatureModel::BooleanConstraint> (C.clone()) );
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        .def(
+            "add_non_boolean_constraint",
+            [](vf::FeatureModelBuilder &FMB, vf::Constraint &C){
+                return FMB.addConstraint(std::make_unique<vf::FeatureModel::NonBooleanConstraint> (C.clone()) );
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "add_non_boolean_constraint",
            [](vf::FeatureModelBuilder &FMB, vf::Constraint &C){
                return FMB.addConstraint(std::make_unique<vf::FeatureModel::NonBooleanConstraint> (C.clone()) );
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        .def(
+            "set_vm_name",
+            [](vf::FeatureModelBuilder &FMB, std::string name){
+                return FMB.setVmName(name);
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "set_vm_name",
            [](vf::FeatureModelBuilder &FMB, std::string name){
                return FMB.setVmName(name);
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            [](vf::FeatureModelBuilder &FMB, std::string name){
+                return FMB.setVmName(name);
+            }    
+        )                
+        .def(
+            "set_path",
+            [](vf::FeatureModelBuilder &FMB, fs::path path){
+                return FMB.setPath(path);
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "set_path",
            [](vf::FeatureModelBuilder &FMB, fs::path path){
                return FMB.setPath(path);
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            [](vf::FeatureModelBuilder &FMB, fs::path path){
+                return FMB.setPath(path);
+            }    
+        )                
+        .def(
+            "set_commit",
+            [](vf::FeatureModelBuilder &FMB, fs::path commit){
+                return FMB.setCommit(commit);
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "set_commit",
            [](vf::FeatureModelBuilder &FMB, fs::path commit){
                return FMB.setCommit(commit);
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            [](vf::FeatureModelBuilder &FMB, fs::path commit){
+                return FMB.setCommit(commit);
+            }    
+        )                
+        .def(
+            "make_root",
+            [](vf::FeatureModelBuilder &FMB, std::string name){
+                return FMB.makeRoot(name);
+            }    
+        )
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]:
```suggestion
        .def(
            "make_root",
            [](vf::FeatureModelBuilder &FMB, std::string name){
                return FMB.makeRoot(name);
            }
        )
```

File: bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp (hunk 3 above)
Comment (vulder) [reply bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp]: Also, please add a bit of documentation to the bindings.
For example:
```
        .def(
            "make_binary_feature",
            [](vf::FeatureModelBuilder &FMB, std::string feature_name, bool Opt){
                return FMB.makeFeature<vf::BinaryFeature>(feature_name); 
            },
            R"pbdoc(Create a new binary `Feature` with the given `feature_name`.)pbdoc"
        )
```

**Review by vulder — CHANGES_REQUESTED** [review]

**Review by vulder — APPROVED** [review]

**Pull request merged**